src/core/net/ip4_types.hpp

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes types and constants for IPv4 processing.
+ */
+
+#ifndef IP4_TYPES_HPP_
```

Comment:
I noticed you followed the same file-names as we have for `ip6` (i.e. have `ip6_types.hpp`, `ip6_address.hpp`, `ip6_headers.hpp` etc).

The reason we have broken the definitions into multiiple headers/files for `ip6` was that the definitions/classes/constants needed to be used from different modules in OT (e.g. many files include `ip6_address.hpp` to get `Ip6::Address` or `Ip6::Prefix` definition). This is mainly to avoid header dependency loops (e.g. the `ip6_types.hpp` was added recently to address such a situation in PR https://github.com/openthread/openthread/pull/7547).

I am not sure if the same situation would be applicable for IPv4 related definitions as they will be all used form NAT64 module only. 

So we can conside simplifying this and have all the IPv4 definitions in one header file/source, e.g. `ip4_types.hpp` and `ip4_types.cpp`. Thoughts?

Reply:
Makes sense.

> I am not sure if the same situation would be applicable for IPv4 related definitions as they will be all used form NAT64 module only.

The possible use case for ipv4 are all related to nat64, and for now the dns client is using the Ip4::Address class.

What makes me a bit worry is after merging these fields, we are indirectly including tcp6 and udp6 when we want to use "Ip4::Address", however, I think we should extract the `Tcp::Header` and `Udp::Header` it really becomes an issue in the future.